### ISSUE TITLE
[Bug]: Property 'returnValue' does not exist on type 'IpcMainInvokeEvent'

### ISSUE THREAD

**Issue by angelhdzmultimedia** [electron/electron#39277]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

25.3.2

### What operating system are you using?

Windows

### Operating System Version

10.0.22621 Build 22621

### What arch are you using?

x64

### Last Known Working Electron version

23.1.3

### Expected Behavior

I have a desktop app I built on March 15 2023.

It has been working great until today while installed.

I expected updating the icon flawlessly, and executing `quasar build -m electron` like the first time without errors.

### Actual Behavior

Today I decided to update the icon. 

Then ran `quasar build -m electron` and got an error related to `yarn` which already reported to Quasar.
But then I checked the `src/electron-main.ts` file because it was red, due to an error.

And the error is `Property 'returnValue' does not exist on type 'IpcMainInvokeEvent'`

Did a search in the Issues for `returnValue` and `IpcMainInvokeEvent.returnValue` and didn't find a duplicate.
And went to the API reference and don't see any mention of this property so apparently it was removed?

### Testcase Gist URL

None

### Additional Information

package.json:

```
{
  "name": "quasar-app",
  "version": "0.0.1",
  "description": "A Quasar project with a panel to execute commands.",
  "productName": "Commands Panel",
  "author": "AngelHdz Multimedia",
  "private": true,
  "scripts": {
    "clean": "rimraf node_modules pnpm-lock.yaml"
  },
  "dependencies": {
    "@electron/remote": "^2.0.10",
    "@quasar/extras": "^1.16.5",
    "pinia": "^2.1.6",
    "quasar": "^2.12.3",
    "unplugin-vue-router": "^0.6.4",
    "vite-plugin-vue-layouts": "^0.8.0",
    "vue": "^3.3.4",
    "vue-router": "^4.2.4"
  },
  "devDependencies": {
    "@quasar/app-vite": "^1.4.3",
    "@types/node": "^12.20.21",
    "autoprefixer": "^10.4.14",
    "electron": "^25.3.2",
    "electron-builder": "^24.6.3",
    "electron-packager": "^17.1.1",
    "typescript": "^4.5.4"
  },
  "engines": {
    "node": "^18 || ^16 || ^14.19",
    "npm": ">= 6.13.4",
    "yarn": ">= 1.21.1"
  }
}
```

**Comment by angelhdzmultimedia** [comment]
### Screenshots

![image](https://github.com/electron/electron/assets/10191085/3922f902-4e7e-4cf7-8526-309cce58d07e)

![image](https://github.com/electron/electron/assets/10191085/e585e778-9b7a-4373-a7c6-5692df636472)


### Reproduction Link

https://github.com/angelhdzmultimedia/quasar-electron-typescript-reprod

**Comment by squalle0nhart** [comment]
Same problem after upgrading to electron 25.x. Do we need to wait electron team to fix it?

**Comment by squalle0nhart** [comment]
I've found the workaround, you can copy electron.d.ts from 24.x to node_modules/electron folder. It will make build success. However it's just temporary workaround.

**Comment by angelhdzmultimedia** [comment]
> I've found the workaround, you can copy electron.d.ts from 24.x to node_modules/electron folder. It will make build success. However it's just temporary workaround.

Thanks will try it.

**Comment by angelhdzmultimedia** [comment]
> I've found the workaround, you can copy electron.d.ts from 24.x to node_modules/electron folder. It will make build success. However it's just temporary workaround.

Sorry to bother you again.

I installed version 24, the error is gone, but it says deprecated. 

![image](https://github.com/electron/electron/assets/10191085/202fa3e5-0a37-4a68-995d-ae786ff5889b)

**Comment by squalle0nhart** [comment]
@angelhdzmultimedia  

I've found the problem. If you have this error try to import Event from Electron 

``` 
import {
 ...
  Event,
  ...
} from 'electron'; 

That'll work

**Comment by squalle0nhart** [comment]
@angelhdzmultimedia 
```
ipcMain.handle('my-invokable-ipc', async (event, ...args) => {
  const result = await somePromise(...args)
  return result
}) 
```
I think you just need to return value, no need event.returnValue

**Comment by angelhdzmultimedia** [comment]
> @angelhdzmultimedia
> 
> ```
> ipcMain.handle('my-invokable-ipc', async (event, ...args) => {
>   const result = await somePromise(...args)
>   return result
> }) 
> ```
> 
> I think you just need to return value, no need event.returnValue

![image](https://github.com/electron/electron/assets/10191085/cc302dd0-abeb-4814-b6a2-473a316f7cb6)

**Comment by angelhdzmultimedia** [comment]
I'm going to try other methods like event.sender, event.reply, or using ipcRenderer.on('some-event', (event, arg) => {
      // Receive the response from main process here
}).

Thanks!

**Comment by MarshallOfSound** [comment]
There are a few different variations of `IpcMain` event types.

`IpcMainEvent` _could_ be a synchronous ipc event so it has a `returnValue` property per the docs for providing the synchronous response

`IpcMainInvokeEvent` is never synchronous and you provide values via a handler function, not via `e.returnValue`.

E.g.

```js
ipcMain.on('my-sync-ipc-msg', (e) => {
  doSomething();
  e.returnValue = true;
});

ipcMain.handle('my-async-ipc-msg', (e) => {
  doSomething();
  return true;
});

// Used in the renderer as
const value = ipcRenderer.sendSync('my-sync-ipc-msg');
const value = await ipcRenderer.invoke('my-async-ipc-msg');
```

**Comment by MafistoPL** [comment]
1. Analysis of the Reported Issue:

* A user encountered an issue related to using `returnValue` within the context of `ipcMain.handle()`. Discussions indicate that in newer versions of Electron, the `ipcMain.handle()` method is intended for handling asynchronous calls from the renderer process and should return values directly, not via `returnValue`.
2. Code Review:

* I searched the project for the use of `.returnValue` and `.handle(` to check if the discussed issue might affect our application.
3. Analysis of a Specific Use Case:

* I found a single instance of `returnValue` usage in the file [IPCAgent.js](https://github.com/postman-eng/postman-app/blob/4cca5ca3add76c12b38ea62d388a6373d59abfdb/src/main/runtime/agents/IPCAgent.js#L469), which was associated with the `subscribe` method in a custom `ipcMain`.
4. Review of the Custom `ipcMain` Implementation:

* I checked the implementations of the `subscribe` method in the custom `ipcMain` across several files. In each case, the `subscribe` method utilizes `ipcMain.on`. This suggests that the changes in the Electron API do not negatively impact our application, as we are using synchronous IPC communication, which is adequately handled by `ipcMain.on`.
